contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -3435,3 +3510,415 @@ objectproperty_internal(PG_FUNCTION_ARGS)
+
+	/* Connect to SPI manager */
+	if ((ret = SPI_connect()) < 0)
+		/* internal error */
+		elog(ERROR, "get_bbf_pivot_tuplestore: SPI_connect returned %d", ret);
+
+	/* Now retrieve the bbf_pivot source rows */
+	ret = SPI_execute_raw_parsetree(sql, true, 0);
+	proc = SPI_processed;
+
+	/* Check for qualifying tuples */
+	if ((ret == SPI_OK_SELECT) && (proc > 0))
+	{
+		SPITupleTable *spi_tuptable = SPI_tuptable;
+		TupleDesc	spi_tupdesc = spi_tuptable->tupdesc;
+		int			ncols = spi_tupdesc->natts;
+		char	   **columngroup;
+		char	   **lastcolumngroup = NULL;
+		bool		firstpass = true;
+		uint64		i;
+		int			j;
+		int			result_ncols;
+
+		if (num_categories == 0)
+		{
+			/* no qualifying category tuples */
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					 errmsg("provided \"categories\" SQL must " \
+							"return 1 column of at least one row")));
+		}
+
+		if (ncols < 2)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("invalid source data SQL statement"),
+					 errdetail("The provided SQL must return 2 " \
+							   " columns; category, and values.")));
+
+		result_ncols = (ncols - 2) + num_categories;
+
+		/* Recheck to make sure we tuple descriptor still looks reasonable */
+		if (tupdesc->natts != result_ncols)
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					 errmsg("invalid return type"),
+					 errdetail("Query-specified return " \
+							   "tuple has %d columns but bbf_pivot " \
+							   "returns %d.", tupdesc->natts, result_ncols)));
+
+		/* allocate space and make sure it's clear */
+		values = (char **) palloc0(result_ncols * sizeof(char *));
+		columngroup = (char **) palloc0((ncols - 2) * sizeof(char *));
+		lastcolumngroup = (char **) palloc0((ncols - 2) * sizeof(char *));
+
+		for (i = 0; i < proc; i++)
+		{
+			HeapTuple	spi_tuple;
+			bbf_pivot_cat_desc *catdesc;
+			char	   *catname;
+			bool 	   	is_new_row = false;
+
+			/* get the next sql result tuple */
+			spi_tuple = spi_tuptable->vals[i];
+
+			if (ncols > 2)
+			{
+				/* get the non-pivot column group from the current sql result tuple */
+				for (j = 0; j < ncols - 2; j++)
+				{	
+					columngroup[j] = SPI_getvalue(spi_tuple, spi_tupdesc, j+1);
+				}
+
+				/*
+				* if we're on a new output row, grab the column values up to
+				* column N-2 now
+				*/
+
+				if (!firstpass)
+				{
+					for (j = 0; j < ncols - 2; j++)
+					{	
+						if (!xstreq(columngroup[j], lastcolumngroup[j]))
+						{
+							is_new_row = true;
+							break;
+						}
+					}
+				}
+
+				if (firstpass || is_new_row)
+				{
+					/*
+					* a new row means we need to flush the old one first, unless
+					* we're on the very first row
+					*/
+					if (!firstpass)
+					{
+						for (j = 0; j < result_ncols; j++)
+						{
+							if (values[j] == NULL)
+								values[j] = pstrdup("0");
+						}
+						/* rowid changed, flush the previous output row */
+						tuple = BuildTupleFromCStrings(attinmeta, values);
```

Comment:
It looks like `BuildTupleFromCStrings` is not doing a memcpy of `values` and you are doing a `xpfree` at ;om 3876 this means that the tuple would be pointing to a freed memory. 

Let's run valgrind here. :)

Reply:
tuple = BuildTupleFromCStrings(attinmeta, values) use values(c string) and attinmeta to build a brand new datum not even in string type, it is more like atoi() (convert string to int) so there is no need to worry about memcpy, and free that value list is necessary later.